File: internal/connectors/fetch_service.go
package connectors

import (
	"elcom/internal/storage"
)

type FetchService struct {
	db        *storage.DB
	connector MailConnector
	store     *MailStoreService
}

type FetchResult struct {
	Fetched int
	Stored  int
}

func NewFetchService(db *storage.DB, rawMailDir string, connector MailConnector) *FetchService {
	return &FetchService{
		db:        db,
		connector: connector,
		store:     NewMailStoreService(db, rawMailDir),
	}
}

func (s *FetchService) FetchAndStore(label string, max int) (FetchResult, error) {
	messages, err := s.connector.FetchInbox(label, max)
	if err != nil {
		return FetchResult{}, err
	}

	stored := 0
	for _, msg := range messages {
		if _, err := s.store.Store(msg); err != nil {
			return FetchResult{}, err
		}
		stored++
	}

	return FetchResult{Fetched: len(messages), Stored: stored}, nil
}

File: src/connectors/imap/fetchService.ts
import { logger } from '../../logger.js';
import { AppDb } from '../../storage/db.js';
import { MailStoreService } from '../mailStoreService.js';
import { ImapConnector } from './imapConnector.js';

export class ImapFetchService {
  private readonly storeService: MailStoreService;

  constructor(
    private readonly db: AppDb,
    private readonly connector: ImapConnector,
  ) {
    this.storeService = new MailStoreService(db);
  }

  async fetchAndStore(label: string, max: number): Promise<{ fetched: number; stored: number }> {
    const fetched = await this.connector.fetchInbox({ label, max });

    let stored = 0;
    for (const msg of fetched) {
      this.storeService.store(msg);
      stored += 1;
    }

    logger.info({ fetched: fetched.length, stored }, 'IMAP fetch completed');
    return { fetched: fetched.length, stored };
  }
}
